### PR TITLE
Static and Dynamic scaling

### PR DESCRIPTION
**Pull request by atayarani**
@marcuswalser @StabbyCutyou @ehealy 
/cc @Tapjoy/eng-group-ops 

Notes: 
* Had to implement a small code change to fix specs.  The error when they failed was misleading.
* Moved the instance counts from args to yaml params
* Added the new gem dependency because it makes CircleCI easier to use

### CODE REVIEW

File: lib/tapjoy/autoscaling_bootstrap.rb [hunk 1]
```diff
@@ -182,8 +186,9 @@ def configure_environment(filename, env=nil)
 
       # Exponential backup
       def aws_wait(tries)
-        puts "Sleeping for #{2 ** tries}..."
-        sleep 2 ** tries
+        backoff_sleep = 2 ** tries
```
Comment (marcuswalser) [inline lib/tapjoy/autoscaling_bootstrap.rb]: Neat exponential backoff.

File: lib/tapjoy/autoscaling_bootstrap.rb [hunk 2]
```diff
@@ -153,7 +154,9 @@ def confirm_config(keypair:, zones:, security_groups:, instance_type:,
 
       # configure environment
 
-      def configure_environment(filename, env=nil)
+      def configure_environment(opts)
+        filename = opts[:filename]
+        env = opts[:env] if opts.key?(:env)
```
Comment (StabbyCutyou) [inline lib/tapjoy/autoscaling_bootstrap.rb]: You could clean this up, I think

```ruby
env = ops[:env] || facet_hash[:environment] || defaults_hash[:environment]
```

File: lib/tapjoy/autoscaling_bootstrap.rb [hunk 3]
```diff
@@ -168,6 +171,7 @@ def configure_environment(filename, env=nil)
 
         new_config = defaults_hash.merge!(env_hash).merge(facet_hash)
         new_config[:config_dir] = config_dir
+        new_config[:instance_ids] = opts[:instance_ids] if opts.key?(:instance_ids)
```
Comment (StabbyCutyou) [inline lib/tapjoy/autoscaling_bootstrap.rb]: if opts[:instance_ids] are emtpty, the value is nil. so not sure what the clll to has_key check is for?

File: lib/tapjoy/autoscaling_bootstrap/AWS/Autoscaling/group.rb [hunk 4]
```diff
@@ -9,11 +9,10 @@ def client
               @client ||= Tapjoy::AutoscalingBootstrap::AWS::Autoscaling.client
             end
 
-            def resize(min_size: 0, max_size: 0, desired_capacity:0)
+            def resize(min: 0, max: 0, desired:0)
               self.client.update_auto_scaling_group(
                 auto_scaling_group_name: Tapjoy::AutoscalingBootstrap.scaler_name,
-                min_size: min_size, max_size: max_size,
-                desired_capacity: desired_capacity)
+                min_size: min, max_size: max, desired_capacity: desired)
```
Comment (StabbyCutyou) [inline lib/tapjoy/autoscaling_bootstrap/AWS/Autoscaling/group.rb]: Should these have been changed to meet their siblings

Comment (atayarani) [inline lib/tapjoy/autoscaling_bootstrap/AWS/Autoscaling/group.rb]: No, since this is a call to AWS and they expect the keys to be like this.

File: lib/tapjoy/autoscaling_bootstrap.rb (hunk 3 above)
Comment (atayarani) [inline lib/tapjoy/autoscaling_bootstrap.rb]: Are you recommending `new_config[:instance_ids] = opts[:instance_ids] if opts[:instance_ids]`?

File: lib/tapjoy/autoscaling_bootstrap/AWS/Autoscaling/group.rb (hunk 4 above)
Comment (StabbyCutyou) [inline lib/tapjoy/autoscaling_bootstrap/AWS/Autoscaling/group.rb]: Gotcha

File: lib/tapjoy/autoscaling_bootstrap.rb (hunk 2 above)
Comment (atayarani) [inline lib/tapjoy/autoscaling_bootstrap.rb]: is that really cleaner?

Comment (StabbyCutyou) [inline lib/tapjoy/autoscaling_bootstrap.rb]: For ruby it is

Comment (StabbyCutyou) [inline lib/tapjoy/autoscaling_bootstrap.rb]: The reason being - now there is ONE place where "env" is set from the multitude of incoming options. Instead of someone coming along later, making a change between this line and one of the others, and having that not be the right place to override something, you have the entire fall-through logic of how to correctly get the env value in one single place.

File: lib/tapjoy/autoscaling_bootstrap.rb (hunk 3 above)
Comment (StabbyCutyou) [inline lib/tapjoy/autoscaling_bootstrap.rb]: In this contxt, you almost don't even care what is in opts[:instance_ids].

If opts[:instance_ids] is nil, new_config[:instance_ids] will still end up nil. So you already must be handling that case downstream somewhere.

Since it doesn't matter if either of them are nil, your check for opts.key? or opts[:instance_id] to conditionally set new_config doesn't matter.

TL;DR;

```ruby
new_config[:instance_ids] = opts[:instance_ids]
```

Same great taste, less code to maintain.

File: bin/tass [hunk 5]
```diff
@@ -94,12 +94,7 @@ when 'scale'
     usage 'scale [options]'
     synopsis 'This command is used to scale up auto scaling groups'
```
Comment (marcuswalser) [inline bin/tass]: And down?

File: lib/tapjoy/autoscaling_bootstrap.rb (hunk 3 above)
Comment (atayarani) [inline lib/tapjoy/autoscaling_bootstrap.rb]: In my original version, new_config[:instance_ids] doesn't get created unless opts[:instance_ids] exists right?

Comment (atayarani) [inline lib/tapjoy/autoscaling_bootstrap.rb]: confirmed, these are not the same thing:

```
irb(main):001:0> test_hash = {a: nil}
=> {:a=>nil}
irb(main):003:0> test_hash.key?(:a)
=> true
irb(main):004:0> test2hash = {b: :test}
=> {:b=>:test}
irb(main):005:0> test2hash[:c] = nil
=> nil
irb(main):006:0> puts test2hash
{:b=>:test, :c=>nil}
=> nil
irb(main):007:0> test2hash[:d] = test_hash[:d] if test_hash.key?(:d)
=> nil
irb(main):008:0> puts test2hash
{:b=>:test, :c=>nil}
=> nil
```

**Comment by marcuswalser** [comment]
My last test of ```tass scale up``` launched a node, but immediately deleted it - I built the gem fresh from the head of this branch and uninstalled tass completely before testing - any ideas?

**Comment by marcuswalser** [comment]
LGTM at this point assuming @StabbyCutyou is satisfied.

**Comment by marcuswalser** [comment]
:clap:
